Self Assessment

aputinski opened the pull request:
#### Description:

Adds a "Survey" option on the dashboard panel.

#### Screenshots:

![Screen Shot 2020-05-07 at 1 25 33 PM](https://user-images.githubusercontent.com/603528/81325481-6abc2300-9066-11ea-83b6-61300698514d.png)
![Screen Shot 2020-05-07 at 1 25 55 PM](https://user-images.githubusercontent.com/603528/81325491-6c85e680-9066-11ea-8691-7ee8343b5a7e.png)

#### How to test:

Press "Survey" and answer the questions.

* You must agree to the first statement to continue
* Selecting "Yes" for "Do you live in a long-term care facility?" will end the survey.
* Selecting "No" for "Are you feeling ill?" will end the survey
* Selecting "Yes" for "Do you have any life-threatening symptoms..." will end the survey.

#### Notes
* The survey will eventually support a server integration. Most of the work for that has been done, but due to time constraints, a few things still need to be cleaned up regarding async loading states / error states.
* The phone number on the "Emergency Services" screen will need to be localized for countries outside of the U.S.
* The local survey is located at `app/views/assessment/survey.en.json`. If a file that corresponds to the user's language is present, it will be loaded, otherwise it will fallback to english.
* All labels are under the top level `assessment` key.

aputinski commented on App.js:
```diff
@@ -16,9 +17,11 @@ const App = () => {
-        <Theme use='default'>
-          <Entry />
-        </Theme>
+        <SafeAreaProvider>
```
For some reasons the app works without this in the current develop branch, but that seems like a bug and was crashing the app locally for me.

tstirrat replied:
Wow. good catch!

aputinski commented on app/constants/apis.js:
```diff
@@ -0,0 +1,8 @@
+export const CATCHA_URL =
+  'https://fathomless-mountain-29102.herokuapp.com/v1.0/captcha';
```
These are all testing URLs and will need to be replaced by each white-labeled app.

aputinski commented on app/views/assessment/Assessment.js:
```diff
@@ -0,0 +1,315 @@
+  );
+  const meta = useMemo(
+    () => ({
+      completeRoute: 'EndComplete',
```
This can be changed to `EndShare` and then the additional share/submit screens will be added to the flow.

Patrick-Erichsen commented:
Couldn't make a comment on the files tab - but can we make a folder inside `images/` for everything related to the survey?

Patrick-Erichsen commented on app/views/assessment/AssessmentEndCaregiver.js:
```diff
@@ -0,0 +1,31 @@
+      description={
+        <Trans i18nKey='assessment.caregiver_description'>
+          <Text />
+          <Text style={{ fontFamily: Fonts.primaryBold }} />
```
`<Typography bold />`

Patrick-Erichsen replied:
Same goes for all of the other `End` files.

Patrick-Erichsen commented on app/views/assessment/AssessmentEndCaregiver.js:
```diff
@@ -0,0 +1,31 @@
+          <Text style={{ fontFamily: Fonts.primaryBold }} />
+        </Trans>
+      }
+      image={require('../../assets/images/illustration-screening-end-caregiver.png')}
```
Same goes for all of the other `End` files.

Patrick-Erichsen commented on app/views/assessment/AssessmentEndDistancing.js:
```diff
@@ -0,0 +1,31 @@
+import React, { useContext } from 'react';
```
Could we make a folder under assessments for all of these end screens? E.g. `app/views/assessment/endScreens/AssessmentEndDistancing.js`

Patrick-Erichsen commented on app/views/assessment/AssessmentEndShare.js:
```diff
@@ -0,0 +1,73 @@
+/** @type {React.FunctionComponent<{}>} */
+const AssessmentEndShare = ({ navigation }) => {
+  let { t } = useTranslation();
+  let [a, setA] = useState(true);
```
can we give this and the line below a name that's more descriptive?

Patrick-Erichsen commented on app/views/assessment/AssessmentOption.js:
```diff
@@ -0,0 +1,181 @@
+  return (
+    <Fragment>
+      <TouchableOpacity
+        onPress={() => {
```
Can we pull this out into a method?

Patrick-Erichsen commented on app/views/assessment/AssessmentOption.js:
```diff
@@ -0,0 +1,181 @@
+        </View>
+      </TouchableOpacity>
+      {showDatePicker && (
+        <DateTimePicker
```
Could you pull this out into `app/components`? Feels generic enough to go in there and then pull into this file. Eventually, I can imagine us wanting to have some custom theming in there, specific language strings, etc.

Patrick-Erichsen replied:
Also - I wasn't able to come across any DateTimePicker elements during my survey.

Patrick-Erichsen commented on app/views/assessment/AssessmentOption.js:
```diff
@@ -0,0 +1,181 @@
+        }}>
+        <View style={[styles.container, selected && styles.containerSelected]}>
+          <View style={styles.primary}>
+            {showIndicator && (
```
Could we pull this into a component? Hard to understand what it is at a glance.

Patrick-Erichsen commented on app/views/assessment/AssessmentOption.js:
```diff
@@ -0,0 +1,181 @@
+  // The API doesn't have a defined way to know a specific option is a date,
+  // we just assume the first option is the date picker when type is SCREEN_TYPE_DATE
+  const isDate = type === SCREEN_TYPE_DATE && index === 0;
+  const showIndicator =
```
I think this is just a type check to make sure that the value of `type` is valid - if so, I think something like `isValidType` would be clearer.

Patrick-Erichsen commented on app/views/assessment/AssessmentOption.js:
```diff
@@ -0,0 +1,181 @@
+}) => {
+  // The API doesn't have a defined way to know a specific option is a date,
+  // we just assume the first option is the date picker when type is SCREEN_TYPE_DATE
+  const isDate = type === SCREEN_TYPE_DATE && index === 0;
```
i think something like `isDateOption` would be clearer here - it took me a bit to realize this isn't related to an actual `Date()` or anything.

Patrick-Erichsen commented on app/views/assessment/AssessmentOption.js:
```diff
@@ -0,0 +1,181 @@
+  index,
+  onSelect,
+  option,
+  selected,
```
i think `isSelected` would be clearer for a boolean val here

Patrick-Erichsen commented on app/views/assessment/AssessmentOption.js:
```diff
@@ -0,0 +1,181 @@
+    (isDate && selected && answer) ? new Date(answer.value) : null,
+  );
+  const label = useMemo(() => {
+    if (isDate && date)
```
please use brackets to and a standard `if/else` to keep this more in line with our style guide.

Patrick-Erichsen commented on app/views/assessment/AssessmentOption.js:
```diff
@@ -0,0 +1,181 @@
+          display='default'
+          is24Hour
+          mode='date'
+          onChange={(e, date) => {
```
Can we pull this out into a method?

Patrick-Erichsen commented on app/views/assessment/AssessmentQuestion.js:
```diff
@@ -0,0 +1,142 @@
+  );
+  /** @type {(value: string, index: number) => void} */
+  const onSelectHandler = (value, index) => {
+    switch (question.question_type) {
```
If there is only one case in this switch I think it makes sense to just use an if/else.

Patrick-Erichsen commented on app/views/assessment/AssessmentQuestion.js:
```diff
@@ -0,0 +1,142 @@
+    switch (question.question_type) {
+      case QUESTION_TYPE_MULTI:
+        setSelectedValues(values => {
+          let exists = values.some(v => v.index === index);
```
What are verifying exists here? It's hard to understand with this variable name.

Also, please use a `const` here since the value is never modified.

Patrick-Erichsen commented on app/views/assessment/AssessmentQuestion.js:
```diff
@@ -0,0 +1,142 @@
+    answers[question.id] || [],
+  );
+  /** @type {(value: string, index: number) => void} */
+  const onSelectHandler = (value, index) => {
```
Could you provide a comment as to what this is doing?  I'm not quite sure.

Patrick-Erichsen commented on app/views/assessment/AssessmentQuestion.js:
```diff
@@ -0,0 +1,142 @@
+          <View style={styles.scrollViewContent}>
+            {question.question_description ? (
+              <View style={styles.descriptionWrapper}>
+                {question.question_description
```
Please pull this out into a method. It's very difficult to understand what this is doing just by looking at the code. It also feels like it can probably be simplified - currently, we are iterating over the same array 3 separate times.

Patrick-Erichsen commented on app/views/assessment/AssessmentQuestion.js:
```diff
@@ -0,0 +1,142 @@
+                  ))}
+              </View>
+            ) : null}
+            {[
```
Please pull this out into a method.

Patrick-Erichsen commented on app/views/assessment/AssessmentStart.js:
```diff
@@ -0,0 +1,121 @@
+        });
+      }}
+      ctaTitle={t('assessment.start_cta')}
+      image={require('../../assets/images/illustration-screening-start.png')}
```
Can we just import this?

Patrick-Erichsen commented on app/views/assessment/AssessmentStart.js:
```diff
@@ -0,0 +1,121 @@
+      }}
+      ctaTitle={t('assessment.start_cta')}
+      image={require('../../assets/images/illustration-screening-start.png')}
+      pretitle={
```
Please pull this out into a separate component.

Patrick-Erichsen commented on app/views/Settings.js:
```diff
@@ -123,6 +123,11 @@ export const SettingsScreen = ({ navigation }) => {
             description={t('share.subtitle')}
             onPress={() => navigation.navigate('ExportScreen')}
+          />
+          <Item
```
This should be wrapped in a Feature Flag for the time being (please see my main comment).

```js
<Feature name='self_assessment'>
  <Item
    label={t('assessment.settings_title')}
    description={t('assessment.subtitle')}
    onPress={() => navigation.navigate('AssessmentScreen')}
     last
   />
</Feature> 
```

You will also need to add an entry to `.env`:

```
flag_self_assessment=false
```

tstirrat commented on app/constants/colors.js:
```diff
@@ -11,6 +12,7 @@ const colors = {
   VIOLET_ALPHA_06: '#f7f8ff', // cannot use transparency
   VIOLET: '#4051DB',
   TRANSPARENT: 'transparent',
+  DARK_ALPHA: 'rgba(0, 0, 0, 0.7)',
```
If we need alpha then its ok to use alpha. but ideally it should be a color defined by UX team and encoded into the `themes.js` so that there is a theme equivalent for the 3 themes.

tstirrat commented on app/constants/colors.js:
```diff
@@ -29,6 +31,14 @@ const colors = {
 
   GRAY_BACKGROUND: '#DADADA',
 
+  ASSESSMENT_BACKGROUND: '#F5F8FC',
```
Agreed, if its really one of colors for the components, fine to go in the component.

But do consider if they're "theme" colors to be reused.

tstirrat commented on app/views/assessment/AssessmentCaptcha.js:
```diff
@@ -0,0 +1,142 @@
+              response: answers[question_key].map(r => r.value),
+            }));
+            // TODO: Loading state / disable button
+            axios
```
This is the only place you use axios. and it seems the usage is simple enough to warrant just using `fetch()` -- there is no complex response parsing, there is no complex body encoding. 

Is axios really needed?

tstirrat commented on app/views/assessment/AssessmentCaptcha.js:
```diff
@@ -0,0 +1,142 @@
+              .post(SURVEY_POST_API, response, {
+                headers: { Authorization: `Basic ${token}` },
+              })
+              .then(response => {
```
Use async/await instead of promises where possible.

tstirrat commented on app/services/CaptchaService.js:
```diff
@@ -0,0 +1,51 @@
+  );
+}
+
+export default HCaptcha;
```
Use named exports instead of default exports here and throughout.

Patrick-Erichsen commented on app/services/CaptchaService.js:
```diff
@@ -0,0 +1,49 @@
+ * @param {*} style: custom style
+ * @param {*} url: base url
+ */
+export const HCaptcha = ({ onMessage, style, uri }) => {
```
So at this point it seems like this is a more generic component than a Captcha service - it's essentially a webview wrapper. 

Do we need to pull the captcha logic out into a distinct component if the only unique part is the uri that's being passed?

Patrick-Erichsen commented on app/views/assessment/Assessment.js:
```diff
@@ -0,0 +1,315 @@
+  /** @type {React.MutableRefObject<SurveyAnswers>} */
+  const answers = useRef({});
+  const survey = useSurvey();
+  // const [survey] = useSurveyAsync();
```
Looks like this wasn't caught in the recent commits.

Patrick-Erichsen commented:
Looks like some good progress has been made on some of those comments but not all of them have been addressed yet - would you guys mind tagging me when this is ready for the next round of review?

aputinski commented:
@Patrick-Erichsen I addressed most of the feedback as well as adding some basic tests. I unfortunately don't have any more time to work on this project. I think it's in a mergeable state and hopefully someone can fix the last few issues to get this in a releasable state.

Patrick-Erichsen commented:
@aputinski understood, and thanks for the context/making those updates.

If it's alright with you I'm going to make a PR back to your branch with some of the recommendations I mentioned. Mostly minor stuff but we really just want to push for some consistent code style guidelines given how many contributors there are on the project

tstirrat commented:
> I think it's in a mergeable state and hopefully someone can fix the last few issues to get this in a releasable state.

We'll get to this soon, we've just had a busy few days

tstirrat commented:
I'm closing this for now, because there is ongoing work to get this ready for merge. That work might end up as a new branch/PR.
